### PR TITLE
AMD's HIP support in RAJA

### PR DESCRIPTION
**Pull request by noelchalmers**
This PR adds support for AMD's HIP backend in RAJA. Currently all unit tests are passing. Tested on ROCm 2.2 with the following cmake command:

```cmake -C ../host-configs/hip.cmake ..```

Remaining issues:
- The compiler outputs several warnings about not being able to vectorize some loops. 
- ~~The ltimes example fails to compile. The issue has be tracked to a serial run in this example.~~

### CODE REVIEW

**Comment by trws** [comment]
From what I see so far, this looks pretty good.  I'm getting success on all tests on multiple systems. My only worries, which are not necessarily merge blockers, are these:

* Building with both HIP and OpenMP does not currently work, is that a fundamental limitation of compiling with hip right now or something we can fix, possibly in a future PR?
* The sum reduction test is somewhat shockingly slow, even compared to the other reduction tests, any intuitions on what might be happening there?

**Comment by trws** [comment]
Ok, I forced it to build HIP with OpenMP and am now more confused than before.  The compilation emitted at least dozens of *errors* but compiled all binaries and passed all tests... Something strange is going on here.  Example excerpt from errors at the bottom. Note that there is a fatal error in atomic_ref, then a missing object file, then a successful link.  Thoughts?



```
In file included from /home/scogland1/Projects/raja/repo/test/unit/test-atomic-ref.cpp:20:
In file included from /home/scogland1/Projects/raja/repo/test/unit/test-atomic-ref.hpp:20:
In file included from /home/scogland1/Projects/raja/repo/include/RAJA/RAJA.hpp:106:
In file included from /home/scogland1/Projects/raja/repo/include/RAJA/util/View.hpp:25:
/home/scogland1/Projects/raja/repo/include/RAJA/pattern/atomic.hpp:229:10: error:  'RAJA::atomic::atomicOr':  no overloaded function has restriction specifiers that are compatible with the ambient context
 'atomicOr'
  return RAJA::atomic::atomicOr(Policy{}, acc, value);
         ^
/home/scogland1/Projects/raja/repo/include/RAJA/pattern/atomic.hpp:246:10: error:  'RAJA::atomic::atomicXor':  no overloaded function has restriction specifiers that are compatible with the ambient contex
t 'atomicXor'
  return RAJA::atomic::atomicXor(Policy{}, acc, value);
         ^
/home/scogland1/Projects/raja/repo/include/RAJA/pattern/atomic.hpp:499:26: note: in instantiation of function template specialization 'RAJA::atomic::atomicXor<RAJA::atomic::omp_atomic, int>' requested her
e
    return RAJA::atomic::atomicXor<Policy>(m_value_ptr, rhs) ^ rhs;
                         ^
/home/scogland1/Projects/raja/repo/test/unit/test-atomic-ref.hpp:152:20: note: in instantiation of member function 'RAJA::atomic::AtomicRef<int, RAJA::atomic::omp_atomic>::operator^=' requested here
    { return other ^= (T)i; }
                   ^
/home/scogland1/Projects/raja/repo/test/unit/test-atomic-ref.hpp:473:15: note: in instantiation of member function 'XorEqOtherOp<int, RAJA::atomic::omp_atomic>::operator()' requested here
      T val = otherop(i);
              ^
/home/scogland1/Projects/raja/repo/test/unit/test-atomic-ref.hpp:539:3: note: in instantiation of function template specialization 'testAtomicRefOther<RAJA::policy::omp::omp_for_exec, RAJA::atomic::omp_at
omic, int, XorEqOtherOp>' requested here
  testAtomicRefOther<ExecPolicy, AtomicPolicy, T, XorEqOtherOp   >(seg, count, list);
  ^
/home/scogland1/Projects/raja/repo/test/unit/test-atomic-ref.hpp:617:3: note: in instantiation of function template specialization 'testAtomicRefIntegral<RAJA::policy::omp::omp_for_exec, RAJA::atomic::omp
_atomic, int, 10000>' requested here
  testAtomicRefIntegral<ExecPolicy, AtomicPolicy, int, 10000>();
  ^
/home/scogland1/Projects/raja/repo/test/unit/test-atomic-ref.cpp:26:3: note: in instantiation of function template specialization 'testAtomicRefPol<RAJA::policy::omp::omp_for_exec, RAJA::atomic::omp_atomi
c>' requested here
  testAtomicRefPol<RAJA::omp_for_exec, RAJA::atomic::omp_atomic>();
  ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
clang-9: error: clang frontend command failed with exit code 70 (use -v to see invocation)
HCC clang version 9.0.0 (/home/jenkins/jenkins-root/workspace/compute-rocm-rel-2.5/external/hcc-tot/clang 7c03a2e9172656a0832bdb8e0a0c9cd2cbf634a1) (/home/jenkins/jenkins-root/workspace/compute-rocm-rel-2
.5/external/hcc-tot/compiler 04d27535fcc4ca8250eaa86e01883c43901cd9ac) (based on HCC 1.3.19212-ef84ec5-7c03a2e-04d2753 )
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/rocm/hcc/bin
clang-9: note: diagnostic msg: PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.
clang-9: note: diagnostic msg:
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang-9: note: diagnostic msg: /tmp/test-atomic-ref-49ed32.cpp
clang-9: note: diagnostic msg: /tmp/test-atomic-ref-adb237.cpp
clang-9: note: diagnostic msg: /tmp/test-atomic-ref-49ed32.sh
clang-9: note: diagnostic msg:

********************
[ 99%] Linking HIP executable ../test-atomic-ref.exe
clang-9: error: no such file or directory: 'CMakeFiles/test-atomic-ref.exe.dir/test-atomic-ref.exe_generated_test-atomic-ref.cpp.o'
[ 99%] Built target test-atomic-ref.exe
```

File: include/RAJA/policy/atomic_builtin.hpp [hunk 1]
```diff
@@ -38,6 +38,7 @@ namespace detail
 
 #if defined(RAJA_COMPILER_MSVC)
 
+RAJA_HOST_DEVICE
```
Comment (MrBurmark) [inline include/RAJA/policy/atomic_builtin.hpp]: Why have these been marked host device?

File: include/RAJA/policy/hip/scan.hpp [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+#include <iterator>
+#include <type_traits>
+
+#if defined(__HCC__)
+#define ROCPRIM_HIP_API 1
+#include "rocprim/device/device_scan.hpp"
+#elif defined(__CUDACC__)
+#include "cub/device/device_scan.cuh"
+#include "cub/util_allocator.cuh"
+#endif
```
Comment (MrBurmark) [inline include/RAJA/policy/hip/scan.hpp]: Why are rocprim and cub possible to use in this file?

File: include/RAJA/policy/sequential/atomic.hpp [hunk 3]
```diff
@@ -33,6 +33,7 @@ struct seq_atomic {
 
 RAJA_SUPPRESS_HD_WARN
 template <typename T>
+RAJA_HOST_DEVICE
```
Comment (MrBurmark) [inline include/RAJA/policy/sequential/atomic.hpp]: Why are these seq_atomic functions marked host device?

**Comment by MrBurmark** [comment]
This looks pretty good to me too.
I'm concerned about maintainability. A lot of the code is the same in the cuda and hip implementations, especially the code for reductions and kernel. It might be a good idea to move and/or abstract some components into a shared directory so that the cuda and hip backends share some of their codebase.

**Comment by rhornung67** [comment]
@MrBurmark that's a good suggestion. We can look into refactoring after this PR gets merged.

**Comment by noelchalmers** [comment]
@trws @MrBurmark I'll try to answer your questions together as I believe they are related. As far as I can tell, AMD's HIP compiler makes a distinction between functions with __host__ and/or __device__ attributes and functions with no attributes at all. I believe that's why some errors get emitted when you try to compile with OpenMP @trws . There's likely a section of the OpenMP code that is in a RAJA_HOST_DEVICE function but is trying to call a function without those specifiers. It may only be missing the __host__ or __device__ version of the function in one of the compilation passes so the linker is still able to make a binary (this is likely bad behavior). This is also the reason that I had to flag several of the atomic functions as RAJA_HOST_DEVICE @MrBurmark

**Comment by noelchalmers** [comment]
@trws Regarding the sum reduction test, I also noticed it was quite slow. It's possible it's using some pinned host memory buffer and continually de-referencing it in a GPU kernel. I'll look into it.

**Comment by trws** [comment]
Interesting, this is something I'd like to track down a little bit.  I
realize that the host/device model for hip is more like the cuda-clang
version than the normal cuda version (and thank goodness for that) and
we are totally ok with going through to fix some of these things if it
will work better.

As to merging HIP and CUDA code paths, I'm for that in the sense of code
maintainability, but with the note that I think we are going to want to
differentiate the backends progressively more going forward (had some
chats with an AMD rep here in Cologne that lead me to believe we may
want to plan for that) so we may want to treat it as premature
optimization at least for the time being.

**Comment by mclarsen** [comment]
Out of curiosity, has anyone tried this new backend out on a MacBook Pro with a Radeon Pro 560X?

**Comment by trws** [comment]
> Out of curiosity, has anyone tried this new backend out on a MacBook Pro with a Radeon Pro 560X?

AFAIK HIP doesn't have OSX support, and I don't have a mac with linux installed bare-metal, so I haven't tried it.

**Comment by artv3** [comment]
I gave this a spin on one our local machines and it worked for me. I'm okay with what we have here, just need to update with develop and fix some conflicts.

**Comment by trws** [comment]
I agree with arturo, LGTM, though I would very much like to see the build at least (if not the tests) integrated into our CI in a new PR.  Does anyone provide a rocm stack docker container we could reasonably leverage for this? For that matter, anybody know a CI with a compatible AMD gpu?

**Comment by davidbeckingsale** [comment]
I can add an entry to the CI. It will the same as our cuda tests though - build only.

**Comment by rhornung67** [comment]
@davidbeckingsale if/when we switch over to azure pipelines, it looks like we may be able to access VMs for AMD GPUs https://azure.microsoft.com/en-us/blog/introducing-nvv4-azure-virtual-machines-for-gpu-visualization-workloads/

**Comment by trws** [comment]
That is really intriguing news @rhornung67, I've heard good things about that service anyway, and it might let us cut down to one.  Definitely something to look into.

**Comment by rhornung67** [comment]
@trws we're using azure pipelines for axom and @davidbeckingsale has CI set up there for Umpire -- very easy to set up testing for linux, windows, and osx in one CI tool. Also, throughput is much better since we wouldn't be competing with other LLNL github projects for limited travis resources.

**Comment by trws** [comment]
@rhornung67 I had no idea, but that's great to hear, I'll have to take a look at the setups over there. Thanks for telling me.

**Comment by davidbeckingsale** [comment]
@noelchalmers  - fixed up Travis and bumped camp to point at master.

**Comment by noelchalmers** [comment]
Thanks @davidbeckingsale Good to see the checks passing now.

File: include/RAJA/policy/atomic_builtin.hpp (hunk 1 above)
Comment (trws) [reply include/RAJA/policy/atomic_builtin.hpp]: Old comment I realize, but IIRC hip offers the builtins directly from clang, so it actually makes some good sense.

**Comment by trws** [comment]
So, this is a *huge* commit, and gives a pretty good view of some places where we should really refactor to make things more generic.  That said, I see no reason to hold this back anymore other than getting it rebased up to current develop.  Are there any outstanding issues we need to address first, or that might make us want to mark it "experimental" or similar?

**Comment by artv3** [comment]
I think it might be good to merge this but note its experimental.

**Comment by davidbeckingsale** [comment]
Thanks everyone!!!!! 🎉

**Pull request merged**